server/drivers/yard2LCD.c

```diff
@@ -0,0 +1,647 @@
+	}
+	
+	y--; //start with 0,0
+	x--;
+	p->framebuf[(y * p->width) + x ] = c;
+}
+
+/*
+ * Returns the number of custom characters
+ */
+MODULE_EXPORT int
+yard_get_free_chars(Driver *drvthis)
+{
+	debug(RPT_DEBUG, "%s: Event 16 - Enter yard_get_free_chars",drvthis->name);
+	return NUM_CUSTOMCHARS;
+}
+
+/*
+ * Sets a custom character
+ */
+MODULE_EXPORT void
+yard_set_char(Driver *drvthis, int n, unsigned char *dat)
+{
+	debug(RPT_DEBUG, "%s: Event 17 - Enter yard_set_char: Nr.%d - %02X",drvthis->name,n,dat);
+
+	if ( (n < 0) || (n >= NUM_CUSTOMCHARS) )
+		return;
+	if (!dat)
+		return;
+
+	yard_hwWriteCGRam(drvthis, n, dat);
+}
+
+/*
+ * Returns the width of a character in pixels
+ */
+MODULE_EXPORT int
+yard_cellwidth(Driver *drvthis)
+{
+	debug(RPT_DEBUG, "%s: Event 18 - Enter yard_cellwidth",drvthis->name);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+
+	return p->cellwidth;
+}
+
+/*
+ * Returns the height of a character in pixels
+ */
+MODULE_EXPORT int
+yard_cellheight(Driver *drvthis)
+{
+	debug(RPT_DEBUG, "%s: Event 19 - Enter yard_cellheight",drvthis->name);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+
+	return p->cellheight;
+}
+
+/*
+ * Reads brightness (in promille)
+ */
+MODULE_EXPORT int
+yard_get_brightness(Driver *drvthis, int state)
+{
+	debug(RPT_DEBUG, "%s: Event 20 - Enter yard_get_brightness: %d",drvthis->name,state);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+
+	return (state == BACKLIGHT_ON) ? p->on_brightness : p->off_brightness;
+}
+
+/*
+ * Sets brightness (in promille)
+ */
+MODULE_EXPORT void
+yard_set_brightness(Driver *drvthis, int state, int promille)
+{
+	debug(RPT_DEBUG, "%s: Event 21 - Enter yard_set_brightness: %d - %d",drvthis->name,state,promille);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+
+	// Error check
+	if (promille < 0 || promille > 1000)
+		return;
+
+	// Store values
+	if (state == BACKLIGHT_ON) {
+		p->on_brightness = promille;
+	}
+	else {
+		p->off_brightness = promille;
+	}
+}
+
+/*
+ * Sets the backlight on or off
+ * The hardware supports any value between 0 and 255
+ */
+MODULE_EXPORT void
+yard_backlight(Driver *drvthis, int on)
+{
+	debug(RPT_DEBUG, "%s: Event 22 - Enter yard_backlight: %d",drvthis->name,on);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+
+	int hardware_value = (on == BACKLIGHT_ON) ? p->on_brightness : p->off_brightness;
+
+	// Map range to hardware [0, 1000] -> [0, 255]
+	hardware_value /= 4;
+	if (hardware_value != p->hw_brightness) 
+	{
+		// Send command and update private data
+		yard_hwSetBrightness(drvthis, hardware_value);
+		p->hw_brightness = (char)hardware_value;
+	}
+}
+
+/*
+ * Provides info about the driver
+ */
+MODULE_EXPORT const char *
+yard_get_info(Driver *drvthis)
+{
+	debug(RPT_DEBUG, "%s: Event 23 - Enter yard_get_info",drvthis->name);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+	
+	strcpy(p->info, "Socket-based driver for Y.A.R.D.2 USB LCD");
+	return p->info;
```

Comment:
Yes the main program is a background server, yard2srvd, which handles the LCD output